applications/nrf_desktop/src/modules/usb_state.c

```diff
@@ -410,7 +420,33 @@ static void report_sent(struct usb_hid_device *usb_hid, struct usb_hid_buf *buf,
 	event->report_id = report_id;
 	event->subscriber = usb_hid;
 	event->error = error;
-	APP_EVENT_SUBMIT(event);
+
+	if (!IS_ENABLED(CONFIG_DESKTOP_USB_HID_REPORT_SENT_ON_SOF)) {
+		APP_EVENT_SUBMIT(event);
+	} else {
+		if (error) {
+			/* Synchronization to USB SOF is not used on send error. Instantly send
+			 * enqueued report waiting for USB SOF to ensure proper HID report sent
+			 * event order.
+			 */
+			report_sent_sof(usb_hid);
+			APP_EVENT_SUBMIT(event);
+		} else {
+			if (!atomic_ptr_cas(&usb_hid->report_sent_on_sof, NULL, event)) {
+				/* Instantly submit previous event to ensure proper HID report sent
+				 * event order.
+				 */
+				LOG_WRN("Missing USB SOF between HID report sent callbacks");
+				struct hid_report_sent_event *prev_event =
+						atomic_ptr_set(&usb_hid->report_sent_on_sof, NULL);
+
+				if (prev_event) {
+					APP_EVENT_SUBMIT(prev_event);
+				}
```

Comment:
Is this check really needed? If we are in this scope it means that the prev_event is not NULL. Otherwise the atomic_ptr_cas(&usb_hid->report_sent_on_sof, NULL, event) would return true.

Reply:
Yes, it's necessary because `report_sent_sof` can be called from an ISR context. Then it's possible that ISR jumps in after `atomic_ptr_cas(&usb_hid->report_sent_on_sof, NULL, event)`, but before `atomic_ptr_set(&usb_hid->report_sent_on_sof, NULL)`. That would lead to `prev_event == NULL` here.
Because of that the `usb_hid->report_sent_on_sof` is initially set to NULL and then it is set to `event` after `prev_event` was actually submitted. Setting `usb_hid->report_sent_on_sof` could result in submitting `event` before `prev_event`

Reply:
Consider separating common code to new function:
```
struct hid_report_sent_event *prev_event = atomic_ptr_set(&usb_hid->report_sent_on_sof, NULL);

if (prev_event) {
	APP_EVENT_SUBMIT(prev_event);
}
```
This is duplicated in `report_sent_sof` function.

Reply:
Actually we can simply call `report_sent_sof` here then. Thanks for pointing this out !

Reply:
Wouldn't this be a bit misleading? It's not sent on sof here.

Reply:
We already do it during error handling and when USB is disabled. I think calling the `report_sent_sof` may be more straightforward than introducing another function for that purpose (we do the same thing as when we handle SOF).

Reply:
Discussed offline with @alstrzebonski. We can keep it as is.